c8d/inspect: Add `Manifests` field

vvoland opened the pull request:
- needs: https://github.com/moby/moby/pull/47526
- needs: https://github.com/moby/moby/pull/48330
- follow up to: https://github.com/moby/moby/pull/47526

Add `Manifests` field to image inspect (`/images/{name}/json`) response.
This is the same as in `/images/json`.


**- How to verify it**
```
curl --unix-socket /var/run/docker.sock http://localhost/images/busybox/json?manifests=1
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
API: `GET /images/{name}/json` response now will return the `Manifests` field containing information about the sub-manifests contained in the image index. This includes things like platform-specific manifests and build attestations.
The new field will only be populated if the request also sets the `manifests` query parameter to `true`.
This acts the same as in the `GET /images/json` endpoint.
WARNING: This is experimental and may change at any time without any backward compatibility.
SDK: Deprecate `client.ImageInspectWithRaw` function in favor of the new `client.ImageInspect`.
```

**- A picture of a cute animal (not mandatory but encouraged)**

laurazard reviewed (approved):
LGTM (took my brain a sec to wrap around the changes 😅 probably because I haven't looked at this in a while)

thaJeztah commented on api/types/image/image_inspect.go:
```diff
@@ -119,4 +119,14 @@ type InspectResponse struct {
+	//
+	// WARNING: This is experimental and may change at any time without any backward
+	// compatibility.
+	Manifests []ManifestSummary `json:"Manifests,omitempty"`
```
This field should probably be gated by API version.

Slightly wondering if this should be (at least currently?) an optional field; similar to how we have the "tree view" optional.

vvoland replied:
Ah right, added a version gate.

Why should it be optional though? You can only inspect one image anyway, so it won't bloat the transmitted data too much.

vvoland commented:
Rebased on top of https://github.com/moby/moby/pull/48330 (as they touch the same code).

thaJeztah commented:
> Why should it be optional though? You can only inspect one image anyway, so it won't bloat the transmitted data too much.

Perhaps less relevant (but perhaps it is?) I was curious if there's a performance impact for getting all the info; but it's very much possible we would need (most of) the data in either case (that said; perhaps there's parts we could keep cached somewhere to save on creating the summary).

But regardless, depending on the number of architectures in an image, and depending on whether there's attestations, the Manifests still produce a lot of extra information. Here's `alpine` as an example;

```bash
docker pull --platform=linux/amd64 alpine

Using default tag: latest
latest: Pulling from library/alpine
c6a83fedfae6: Download complete
Digest: sha256:0a4eaa0eecf5f8c050e5bba433f58c052be7587ee8af3e8b3910ef9ab5fbe9f5
Status: Downloaded newer image for alpine:latest
docker.io/library/alpine:latest
```

```bash
docker image inspect alpine
```

This outputs;

- A summary of the default platform, or if missing the first available platform
- A summary of all associated manifests (images, attestations); also for those that are not currently pulled (`Available: false`);

```json
[
    {
        "Id": "sha256:0a4eaa0eecf5f8c050e5bba433f58c052be7587ee8af3e8b3910ef9ab5fbe9f5",
        "RepoTags": [
            "alpine:latest"
        ],
        "RepoDigests": [
            "alpine@sha256:0a4eaa0eecf5f8c050e5bba433f58c052be7587ee8af3e8b3910ef9ab5fbe9f5"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "2024-07-22T22:26:43.778747613Z",
        "DockerVersion": "",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "/bin/sh"
            ],
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 2381,
        "GraphDriver": {
            "Data": null,
            "Name": "overlayfs"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:78561cef0761903dd2f7d09856150a6d4fb48967a8f113f3e33d79effbf59a07"
            ]
        },
        "Metadata": {
            "LastTagTime": "2024-08-15T09:43:26.820642347Z"
        },
        "Manifests": [
            {
                "ID": "sha256:eddacbc7e24bf8799a4ed3cdcfa50d4b88a323695ad80f317b6629883b2c2a78",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:eddacbc7e24bf8799a4ed3cdcfa50d4b88a323695ad80f317b6629883b2c2a78",
                    "size": 528,
                    "platform": {
                        "architecture": "amd64",
                        "os": "linux"
                    }
                },
                "Available": true,
                "Size": {
                    "Content": 3624891,
                    "Total": 12083131
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "amd64",
                        "os": "linux"
                    },
                    "Size": {
                        "Unpacked": 8458240
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:5c7e326e3c8a8c51654a6c5d94dac98d7f6fc4b2a762d86aaf67b7e76a6aee46",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:5c7e326e3c8a8c51654a6c5d94dac98d7f6fc4b2a762d86aaf67b7e76a6aee46",
                    "size": 528,
                    "platform": {
                        "architecture": "arm",
                        "os": "linux",
                        "variant": "v6"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "arm",
                        "os": "linux",
                        "variant": "v6"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:fda9b1b812b25c68f94da5b719039bfa9a3b76e167a8f87e7fc62cb159d21ca1",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:fda9b1b812b25c68f94da5b719039bfa9a3b76e167a8f87e7fc62cb159d21ca1",
                    "size": 528,
                    "platform": {
                        "architecture": "arm",
                        "os": "linux",
                        "variant": "v7"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "arm",
                        "os": "linux",
                        "variant": "v7"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:24ba417e25e780ff13c888ccb1badec5b027944666ff695681909bafe09a3944",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:24ba417e25e780ff13c888ccb1badec5b027944666ff695681909bafe09a3944",
                    "size": 528,
                    "platform": {
                        "architecture": "arm64",
                        "os": "linux",
                        "variant": "v8"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "arm64",
                        "os": "linux",
                        "variant": "v8"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:fa66aa594ffa884dff44f4a97821756545834505df611c375a30c45926402f89",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:fa66aa594ffa884dff44f4a97821756545834505df611c375a30c45926402f89",
                    "size": 528,
                    "platform": {
                        "architecture": "386",
                        "os": "linux"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "386",
                        "os": "linux"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:a01843eb870e11bb20c78a9068269c810f14dd5c49364064fa3f9cf798f666dd",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:a01843eb870e11bb20c78a9068269c810f14dd5c49364064fa3f9cf798f666dd",
                    "size": 528,
                    "platform": {
                        "architecture": "ppc64le",
                        "os": "linux"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "ppc64le",
                        "os": "linux"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:e99a4d9aa9f905cee4171c6d616e4008fa32202fa8aa8aa65efcafbc3a0f5fa5",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:e99a4d9aa9f905cee4171c6d616e4008fa32202fa8aa8aa65efcafbc3a0f5fa5",
                    "size": 528,
                    "platform": {
                        "architecture": "riscv64",
                        "os": "linux"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "riscv64",
                        "os": "linux"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            },
            {
                "ID": "sha256:14da06d3a8959002fd621dd3994a254e2126d239f2fe69e829fd95d16ce81dea",
                "Descriptor": {
                    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                    "digest": "sha256:14da06d3a8959002fd621dd3994a254e2126d239f2fe69e829fd95d16ce81dea",
                    "size": 528,
                    "platform": {
                        "architecture": "s390x",
                        "os": "linux"
                    }
                },
                "Available": false,
                "Size": {
                    "Content": 0,
                    "Total": 0
                },
                "Kind": "image",
                "ImageData": {
                    "Platform": {
                        "architecture": "s390x",
                        "os": "linux"
                    },
                    "Size": {
                        "Unpacked": 0
                    },
                    "Containers": null
                }
            }
        ]
    }
]
```

That data is "per request", but it's possible to inspect multiple images; that results in multiple requests, but the output on the client side will be a combination of all of those;

```bash
docker tag alpine blpine

docker image inspect alpine blpine
```


Some of that for sure is a limitation on the client-side; dumping the JSON output will always be verbose, and not very user-friendly (we need to start working on the "human friendly" versions of `docker image inspect` (`docker image info` ?)). But even with that, I'm wondering if the in-depth information is always relevant; more so if not all platforms were pulled.

So I'm wondering if we need to make these "opt-in" to have more granularity;

- default could continue to be "only a summary of the default (or if missing: first) platform"
- optional with the nested ones
- perhaps some option to show/hide "non-available" ones

We also need to look at a `--platform` flag for this; not exactly sure (yet) what we want that to act like;

- At least it can control "which platform" to produce the overall summary for (as the image _config_ is only present for the overall summary, and may have relevant differences between platforms)
- Do we want it to act as a filter for the nested manifests? Do we need a separate option for that (`manifests=<all|available|platforms=....`)?

vvoland commented:
> Perhaps less relevant (but perhaps it is?) I was curious if there's a performance impact for getting all the info; but it's very much possible we would need (most of) the data in either case (that said; perhaps there's parts we could keep cached somewhere to save on creating the summary).

We're already traversing all of these (to pick the "best representative" variant and calculate the size), so the only performance impact here is just storing the data in the JSON.

> So I'm wondering if we need to make these "opt-in" to have more granularity;

IMO, opt-in for `inspect` is awkward because:
- In most cases you use `docker inspect` directly
- The extra `manifests` or `platform` opt-in only make sense for the `image inspect`
- So we would either have to add these to the `docker inspect` (which would only for for `image`) or only to `docker image inspect` (which would make it impossible to use for `docker inspect`)

Also, `docker inspect` purpose was always to "return low-level information on Docker objects" - I think it's expected to have a non-human friendly output.

>  (we need to start working on the "human friendly" versions of docker image inspect (docker image info ?)).

And this is the direction I'd like to focus next - we really need that (and not just for images, but all "objects" in general). That's where we can have all the granularity we want and provide a better human-readable output, while `docker inspect` would still remain a "plumbing" command.

vvoland committed:
c8d/list: multiPlatformSummary

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/inspect: Avoid interim images.Image conversion

Don't use the `GetImage` call which returns a "best-effort" view of the
image that is compatible with the old images.Image response.

Instead, use the multi-platform view of the image to construct the
inspect response.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Updated; the field will now only be returned if the `manifests` query parameter is set.

vvoland commented on client/image_inspect.go:
```diff
@@ -5,28 +5,80 @@ import (
 
-// ImageInspectWithRaw returns the image information and its raw representation.
-func (cli *Client) ImageInspectWithRaw(ctx context.Context, imageID string) (image.InspectResponse, []byte, error) {
+type imageInspectClientOpt struct {
+	raw        *bytes.Buffer
+	apiOptions image.InspectOptions
+}
+
+type ImageInspectOpt func(*imageInspectClientOpt)
+
+// ImageInspectWithRawResponse instructs the client to additionally store the
+// raw inspect response in the provided buffer.
+func ImageInspectWithRawResponse(raw *bytes.Buffer) ImageInspectOpt {
+	return func(opts *imageInspectClientOpt) {
+		opts.raw = raw
+	}
+}
+
+// ImageInspectWithOpts sets the API options for the image inspect operation.
+func ImageInspectWithOpts(opts image.InspectOptions) ImageInspectOpt {
+	return func(clientOpts *imageInspectClientOpt) {
+		clientOpts.apiOptions = opts
+	}
+}
+
+// ImageInspect returns the image information.
+func (cli *Client) ImageInspect(ctx context.Context, imageID string, inspectOpts ...ImageInspectOpt) (image.InspectResponse, error) {
```
I can't say I fully like it. The "raw response" part is strictly a client feature, so it doesn't make sense to put it directly in the `image.InspectOptions`, on the other hand that forces us to have a client specific opts..

Any thoughts on this one? Would be nice to think this carefully, as it breaks the client interface.

vvoland replied:
The linter will complain about the usage of deprecated `ImageInspectWithRaw`, but for now I'll refrain from adjusting these in case we reach a different conclusion on the shape of this.

thaJeztah replied:
Hm;  I'll have a look at this one!

(For the linter; if there's  too many places to suppress it,  we  can also add a global exception in `.golangci-lint.yaml`)

vvoland replied:
Actually I can just break the `Deprecated` comment for now to make the linter happy 😅

thaJeztah commented on api/swagger.yaml:
```diff
@@ -2004,6 +2004,20 @@ definitions:
+            provides a more detailed view of the platform-specific image manifests or
+            other image-attached data like build attestations.
+
+            Only available if the daemon provides a multi-platform image store.
```
... And if the  `manifests` option  is  set?

thaJeztah commented on api/swagger.yaml:
```diff
@@ -2004,6 +2004,20 @@ definitions:
         $ref: "#/definitions/OCIDescriptor"
+      Manifests:
+        description: |
+            Manifests is a list of image manifests available in this image.  It
```
I think there's a  double space here before `It`

vvoland replied:
Yep, thanks

thaJeztah commented on client/image_inspect.go:
```diff
@@ -5,28 +5,80 @@ import (
+	var buf bytes.Buffer
+	resp, err := cli.ImageInspect(ctx, imageID, ImageInspectWithRawResponse(&buf))
+	if err != nil {
+		return resp, nil, err
```
Nit; better to be  explicit that we're not returning anything here (yeah, we're doing that in many  places,  but as  this is modified in this PR);

```suggestion
		return image.InspectResponse{}, nil, err
```

thaJeztah reviewed (commented):
Thanks for updating; only  had  a quick glance over (just some quick comments, but  didn't  do  a  full review)

laurazard reviewed (approved):
LGTM

thaJeztah commented on daemon/containerd/image_inspect.go:
```diff
@@ -134,6 +122,51 @@ func (i *ImageService) ImageInspect(ctx context.Context, refOrID string, _ backe
+		if err != nil {
+			log.G(ctx).WithField("name", name).WithError(err).Error("failed to parse image name as reference")
+			// Include the malformed name in RepoTags to be consistent with `docker image ls`.
+			repoTags = append(repoTags, img.Name)
```
Pretty unrelated; I thought for a bit we still had a bug to fix around this;

- https://github.com/moby/moby/issues/48747


But we fixed that "in post";

- https://github.com/moby/moby/pull/48777

Wondering now if we should look at handling de-duplicating here (in a follow-up); i.e., if it would be more efficient to do it right at the start, WDYT?

vvoland replied:
Right, let me update

thaJeztah replied:
It's fine for a follow up as it's not directly related! Was reviewing this PR, and I think (almost done) it looks good otherwise

vvoland replied:
Was wondering about the performance difference between collecting them to map & then copying keys to slices between collecting to slice and sorting it afterwards.
Results:
```
BenchmarkCollectRepoTagsAndDigests
BenchmarkCollectRepoTagsAndDigests/slice
BenchmarkCollectRepoTagsAndDigests/slice-8         	 627352	    18742 ns/op	   5644 B/op	    118 allocs/op
BenchmarkCollectRepoTagsAndDigests/map
BenchmarkCollectRepoTagsAndDigests/map-8           	 635817	    18702 ns/op	   5397 B/op	    116 allocs/op
PASS
```

<details>

<summary>Benchmark code</summary>

```go
package containerd

import (
	"context"
	"testing"

	c8dimages "github.com/containerd/containerd/v2/core/images"
	"github.com/containerd/log"
	"github.com/distribution/reference"
	"github.com/docker/docker/internal/sliceutil"
	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
)

func BenchmarkCollectRepoTagsAndDigests(b *testing.B) {
	ctx := context.Background()
	tagged := []c8dimages.Image{
		{Name: "docker.io/library/alpine:latest", Target: ocispec.Descriptor{Digest: "sha256:20e6ec8bc89f4bd24469283fe6c147cfed1ad837e435386ed050d8c57c58eb91"}},
		{Name: "docker.io/library/alpine:3.14", Target: ocispec.Descriptor{Digest: "sha256:30e6ec8bc89f4bd24469283fe6c147cfed1ad837e435386ed050d8c57c58eb92"}},
		{Name: "docker.io/library/alpine:3.13", Target: ocispec.Descriptor{Digest: "sha256:40e6ec8bc89f4bd24469283fe6c147cfed1ad837e435386ed050d8c57c58eb93"}},
		{Name: "docker.io/library/alpine:3.13", Target: ocispec.Descriptor{Digest: "sha256:40e6ec8bc89f4bd24469283fe6c147cfed1ad837e435386ed050d8c57c58eb93"}},
		{Name: "docker.io/library/alpine@sha256:20e6ec8bc89f4bd24469283fe6c147cfed1ad837e435386ed050d8c57c58eb91", Target: ocispec.Descriptor{Digest: "sha256:20e6ec8bc89f4bd24469283fe6c147cfed1ad837e435386ed050d8c57c58eb91"}},
		{Name: "docker.io/library/alpine@sha256:21a3deaa0d32a8057914f36584b5288d2e5ecc984380bc0118285c70fa8c9300", Target: ocispec.Descriptor{Digest: "sha256:21a3deaa0d32a8057914f36584b5288d2e5ecc984380bc0118285c70fa8c9300"}},
	}

	b.Run("slice", func(b *testing.B) {
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			collectRepoTagsAndDigests(ctx, tagged)
		}
		b.ReportAllocs()
	})

	b.Run("map", func(b *testing.B) {
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			collectRepoTagsAndDigestsWithMap(ctx, tagged)
		}
		b.ReportAllocs()

	})
}

func collectRepoTagsAndDigests(ctx context.Context, tagged []c8dimages.Image) (repoTags []string, repoDigests []string) {
	repoTags = make([]string, 0, len(tagged))
	repoDigests = make([]string, 0, len(tagged))
	for _, img := range tagged {
		if isDanglingImage(img) {
			if len(tagged) > 1 {
				// This is unexpected - dangling image should be deleted
				// as soon as another image with the same target is created.
				// Log a warning, but don't error out the whole operation.
				log.G(ctx).WithField("refs", tagged).Warn("multiple images have the same target, but one of them is still dangling")
			}
			continue
		}

		name, err := reference.ParseNamed(img.Name)
		if err != nil {
			log.G(ctx).WithField("name", name).WithError(err).Error("failed to parse image name as reference")
			// Include the malformed name in RepoTags to be consistent with `docker image ls`.
			repoTags = append(repoTags, img.Name)
			continue
		}

		repoTags = append(repoTags, reference.FamiliarString(name))
		if _, ok := name.(reference.Digested); ok {
			repoDigests = append(repoDigests, reference.FamiliarString(name))
			// Image name is a digested reference already, so no need to create a digested reference.
			continue
		}

		digested, err := reference.WithDigest(reference.TrimNamed(name), img.Target.Digest)
		if err != nil {
			// This could only happen if digest is invalid, but considering that
			// we get it from the Descriptor it's highly unlikely.
			// Log error just in case.
			log.G(ctx).WithError(err).Error("failed to create digested reference")
			continue
		}
		repoDigests = append(repoDigests, reference.FamiliarString(digested))
	}
	return sliceutil.Dedup(repoTags), sliceutil.Dedup(repoDigests)
}

// collectRepoTagsAndDigestsWithMap returns repoTags and repoDigests for images with the same target using maps for deduplication.
func collectRepoTagsAndDigestsWithMap(ctx context.Context, tagged []c8dimages.Image) (repoTags []string, repoDigests []string) {
	repoTagsMap := make(map[string]struct{})
	repoDigestsMap := make(map[string]struct{})

	for _, img := range tagged {
		if isDanglingImage(img) {
			if len(tagged) > 1 {
				// This is unexpected - dangling image should be deleted
				// as soon as another image with the same target is created.
				// Log a warning, but don't error out the whole operation.
				log.G(ctx).WithField("refs", tagged).Warn("multiple images have the same target, but one of them is still dangling")
			}
			continue
		}

		name, err := reference.ParseNamed(img.Name)
		if err != nil {
			log.G(ctx).WithField("name", name).WithError(err).Error("failed to parse image name as reference")
			// Include the malformed name in RepoTags to be consistent with `docker image ls`.
			repoTagsMap[img.Name] = struct{}{}
			continue
		}

		repoTagsMap[reference.FamiliarString(name)] = struct{}{}
		if _, ok := name.(reference.Digested); ok {
			repoDigestsMap[reference.FamiliarString(name)] = struct{}{}
			// Image name is a digested reference already, so no need to create a digested reference.
			continue
		}

		digested, err := reference.WithDigest(reference.TrimNamed(name), img.Target.Digest)
		if err != nil {
			// This could only happen if digest is invalid, but considering that
			// we get it from the Descriptor it's highly unlikely.
			// Log error just in case.
			log.G(ctx).WithError(err).Error("failed to create digested reference")
			continue
		}
		repoDigestsMap[reference.FamiliarString(digested)] = struct{}{}
	}

	repoTags = make([]string, 0, len(repoTagsMap))
	repoDigests = make([]string, 0, len(repoDigestsMap))
	// Convert maps to slices
	for tag := range repoTagsMap {
		repoTags = append(repoTags, tag)
	}
	for digest := range repoDigestsMap {
		repoDigests = append(repoDigests, digest)
	}
	return repoTags, repoDigests
}

```
</details>

So not a big difference, slice is slightly faster but does more a bit more allocs than the map.

I'll go with map.

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -332,7 +332,17 @@ func (ir *imageRouter) deleteImages(ctx context.Context, w http.ResponseWriter,
+		return err
+	}
+
+	if r.Form.Get("manifests") != "" && versions.LessThan(httputils.VersionFromContext(ctx), "1.48") {
+		return errdefs.InvalidParameter(errors.New("manifests parameter is not supported for API version < 1.48"))
+	}
```
It's probably fine to have checks on both clients and API. I was considering that this is an option that we could consider to be ignored on the daemon side, i.e., I _think_ the client would be able to gracefully handle responses where this information is not present.


No need to change, just as a comment, in case we want to revisit this.

vvoland replied:
Hmm right, older daemons that don't understand the new API wouldn't complain about this, so I think there's no point to make newer daemon fail in this case?

thaJeztah commented on api/types/image/opts.go:
```diff
@@ -103,6 +103,11 @@ type LoadOptions struct {
 	Platforms []ocispec.Platform
 }
 
+type InspectOptions struct {
+	// Manifests returns the image manifests.
+	Manifests bool
+}
+
```
It just occurred to me that there's 2 option structs now;

- `backend.ImageInspectOptions` (used by the API server)
- `image.InspectOptions` (used by the client)


~The last one, by design, is not exported; it's made internal through `imageInspectClientOpt.apiOptions` (non-exported type AND non-exported field)~

~Given that those are not exported currently, perhaps we should (for now at least) move  this type to the client, and un-export it?~

:warning: 👉 edit: ignore me; there's `ImageInspectWithOpts`, which allows all options to be passed as a struct

thaJeztah commented on client/image_inspect.go:
```diff
@@ -5,28 +5,80 @@ import (
 
-// ImageInspectWithRaw returns the image information and its raw representation.
-func (cli *Client) ImageInspectWithRaw(ctx context.Context, imageID string) (image.InspectResponse, []byte, error) {
+type imageInspectClientOpt struct {
+	raw        *bytes.Buffer
+	apiOptions image.InspectOptions
+}
+
+type ImageInspectOpt func(*imageInspectClientOpt)
```
This part is OK for now, and I like the idea of non-exported fields for "client" and "api" options for clearer separation.

The only thing here that I want to look into a bit (not for this PR) is what design pattern is most logical; the confusing bit here is that `ImageInspectOpt` is an exported type, but it includes a non-exported type in its signature. That means that no implementations can exist outside of those implemented in this package (which currently is fine?).

Mostly wondering if the non-exported type in its signature could be confusing, and it it would make sense to (at some point) export the `imageInspectClientOpt` type itself, but keep its fields (those that we don't want to be set through other means than functional arguments) un-exported.


LOL; hope I'm making sense here!

vvoland replied:
Yeah so actually I wanted to go one step further and make the `ImageInspectOpt` itself unexported, because with all `imageInspectClientOpt` fields being private, there's no benefit for the user to be able to construct its own opt func.

That didn't work though, because linter was complaining about the exported functions using the unexported type.

Alternatively.. we could drop the func opts and just have `ImageInspectOpt` that would embed `image.InspectOptions` and exported `Raw`.

vvoland replied:
Added a commit with alternative approach: https://github.com/moby/moby/pull/48264/commits/51bfa52ac14cdd4754613ecaf1181dd7f83f20b2

Thoughts? @thaJeztah @laurazard

cpuguy83 replied:
Buildkit uses interfaces for option types which makes it much easier to discover what can go there.

cpuguy83 replied:
So it could be something like

```go
type ImageOption interface {
  SetImageOption(*ImageOpt)
}
```

Then you can do something like

```go
type ImageOptFunc func(*ImageOpt)

func(f ImageOptFunc) SetImageOption(o *ImageOpt) {
  f(o)
}
```

vvoland replied:
Is this what you had in mind https://github.com/moby/moby/pull/48264/commits/6d852cc79040f53b1dbf1cb7297cb794a87959d0?

Tbh, I don't see how it makes the options more discoverable, but maybe it's a matter of the IDE, so I don't mind it either. (or I just got your suggestion wrong 😅)

thaJeztah replied:
(Ugh, I thought I posted this comment earlier, but looks like I didn't)


I'm also slightly confused; can you explain more?

Looking at these minimal examples;

<details>

```go
package godoc

type ImageOpt struct {
	something, otherThing string
	feature               bool
}

type ImageOption func(*ImageOpt)

func WithSomething(value string) ImageOption {
	return func(opt *ImageOpt) {
		opt.something = value
	}
}

func WithOtherThing(value string) ImageOption {
	return func(opt *ImageOpt) {
		opt.otherThing = value
	}
}

func WithEnableFeature() ImageOption {
	return func(opt *ImageOpt) {
		opt.feature = true
	}
}
```

</details>


<details>

```go
package godoc2

type ImageOpt struct {
	something, otherThing string
	feature               bool
}

type ImageOption interface {
	Apply(*ImageOpt)
}

type ImageOptFunc func(*ImageOpt)

func (f ImageOptFunc) Apply(o *ImageOpt) {
	f(o)
}

func WithSomething(value string) ImageOption {
	return ImageOptFunc(func(opt *ImageOpt) {
		opt.something = value
	})
}

func WithOtherThing(value string) ImageOption {
	return ImageOptFunc(func(opt *ImageOpt) {
		opt.otherThing = value
	})
}

func WithEnableFeature() ImageOption {
	return ImageOptFunc(func(opt *ImageOpt) {
		opt.feature = true
	})
}
```

</details>


Then GoDoc at least shows all options under the `ImageInspectOpt` type, but I think _both_ do that the same?


<img width="1283" alt="Screenshot 2025-02-04 at 18 58 29" src="https://github.com/user-attachments/assets/35a7f5cd-1e29-4c22-928e-81f9e4da7cb6" />
<img width="1283" alt="Screenshot 2025-02-04 at 19 02 26" src="https://github.com/user-attachments/assets/1717e1a2-c208-43b3-9d8a-dafea12ecd47" />

One exception would be, when changing any of those to return `func(*ImageOpt)` (instead of explictly `ImageOption`), which would still satisfy the type, but won't show it as such in GoDoc, and I _think_ that would be the same for both?

vvoland replied:
@cpuguy83  PTAL

thaJeztah commented on client/image_inspect.go:
```diff
@@ -5,28 +5,80 @@ import (
-	body, err := io.ReadAll(serverResp.body)
-	if err != nil {
-		return image.InspectResponse{}, nil, err
+	buf := opts.raw
+	if buf == nil {
+		buf = &bytes.Buffer{}
+	}
+
+	if _, err := io.Copy(buf, serverResp.body); err != nil {
+		return image.InspectResponse{}, err
```
Would this require some `io.TeeReader`? i.e., if I pass my own custom reader (effectively what `ImageInspectWithRawResponse` does), would interacting with that reader cause things to go bad?

If this still needs some thinking, we could consider un-exporting `ImageInspectWithRawResponse` in the meantime (and only use it internally for the time being)

vvoland replied:
Ah, good point, we can actually just use the `io.TeeReader` and also allow the client to pass any `io.Writer` instead of `bytes.Buffer` specifically.
Let me update.

thaJeztah reviewed (approved):
LGTM


one question on the "raw" custom reader passed, but maybe I'm thinking wrong there (otherwise, we could consider un-exporting it in the meantime)

vvoland committed:
all: Replace deprecated ImageInspectWithRaw usage

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
client/image-inspect: Introduce client opts

Deprecate ImageInspectWithRaw and add a simpler ImageInspect function
which takes optional options.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/inspect: Add Manifests field

Add `Manifests` field to image inspect (`/images/{name}/json`) response.
This is the same as in `/images/json`.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Rebased

thaJeztah commented on client/client_interfaces.go:
```diff
@@ -115,7 +115,9 @@ type ImageAPIClient interface {
 	ImageCreate(ctx context.Context, parentReference string, options image.CreateOptions) (io.ReadCloser, error)
 	ImageHistory(ctx context.Context, image string, opts image.HistoryOptions) ([]image.HistoryResponseItem, error)
 	ImageImport(ctx context.Context, source image.ImportSource, ref string, options image.ImportOptions) (io.ReadCloser, error)
+	// Deprecated: Use [Client.ImageInspect] instead.
```
Perhaps add a hint to use it with the `[client. ImageInspectWithRawResponse]` to read the raw response.

vvoland replied:
Done

thaJeztah commented on daemon/containerd/image_inspect.go:
```diff
@@ -162,7 +165,16 @@ func (i *ImageService) collectRepoTagsAndDigests(ctx context.Context, tagged []c
+	}
+	for digest := range repoDigestsMap {
+		repoDigests = append(repoDigests, digest)
 	}
```
As we're using a map as intermediate, do we need to sort these afterwards to keep a stable result?

vvoland replied:
Hmmm, right. In this case I think it's better to just go with the slice-based implementation: https://github.com/moby/moby/pull/48264#discussion_r1934218435

thaJeztah replied:
Maybe? Are the values sorted when we receive them, or are they equally random? If we go for the slices, we could use the utility we have; https://github.com/moby/moby/blob/02563f2805a7d8a3078196d608e5e8fe08c71671/internal/sliceutil/sliceutil.go#L6-L16

vvoland replied:
They're not really sorted AFAIK, but also not just random. I think it's good to preserve the order.

vvoland replied:
Yep the slices approach also does use the `sliceutil.Dedup` at the end

thaJeztah replied:
Possibly can use the `Dedup` utility that we implemented here; https://github.com/moby/moby/blob/02563f2805a7d8a3078196d608e5e8fe08c71671/internal/sliceutil/sliceutil.go#L6-L16

thaJeztah reviewed (commented):
couple of small comments, but LGTM otherwise

perhaps pending the discussion on the options signature, but I think we can still changes that one as the way they're used won't change.

vvoland commented on client/image_inspect.go:
```diff
@@ -4,29 +4,88 @@ import (
+	apiOptions image.InspectOptions
+}
+
+type ImageInspectOptionFunc func(opt *imageInspectOpts) error
+
+func (f ImageInspectOptionFunc) Apply(o *imageInspectOpts) error {
+	return f(o)
+}
```
Yep, we can keep the `ImageInspectOptionFunc` un-exported.

vvoland commented on client/image_inspect.go:
```diff
@@ -4,29 +4,88 @@ import (
+	})
+}
+
+type ImageInspectOption interface {
+	Apply(*imageInspectOpts) error
+}
```
I wasn't sure on this one. On one hand it's public so it makes sense to put it on top, on the other it's only expected to be implemented by this package, so it's not useful to have it on top.
But I'm fine either way.

thaJeztah commented:
oh! probably new tests were added in the "mount from image" PR;

```
integration/volume/mount_test.go:198:18: SA1019: apiClient.ImageInspectWithRaw is deprecated: Use [Client.ImageInspect] instead. Raw response can be obtained by [ImageInspectWithRawResponse] option. (staticcheck)
				img, _, _ := apiClient.ImageInspectWithRaw(ctx, testImage)
				             ^
integration/volume/mount_test.go:351:14: SA1019: client.ImageInspectWithRaw is deprecated: Use [Client.ImageInspect] instead. Raw response can be obtained by [ImageInspectWithRawResponse] option. (staticcheck)
	_, _, err = client.ImageInspectWithRaw(ctx, imgName)
	            ^
```

vvoland commented:
Right! Fixed

vvoland commented:
Green, I'm going to get this one in. We can still change the interface before the v28 release if we need.
